src/v/cluster/topics_frontend.cc

```diff
@@ -949,4 +982,137 @@ topics_frontend::do_cancel_moving_partition_replicas(
     co_return results;
 }
 
+ss::future<std::error_code> topics_frontend::change_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor_type new_replication_factor,
+  model::timeout_clock::time_point timeout) {
+    auto tp_metadata = _topics.local().get_topic_metadata_ref(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+    if (!tp_metadata.value().get().is_topic_replicable()) {
+        co_return errc::topic_operation_error;
+    }
+
+    auto current_replication_factor
+      = tp_metadata.value().get().get_replication_factor();
+
+    if (current_replication_factor == new_replication_factor) {
+        co_return errc::success;
+    }
+
+    new_replicas_assigment new_replicas;
+
+    // We should hold allocation_units from reallocate_partition while we do
+    // replicate_and_wait.
+    std::optional<allocation_units> allocation_units;
+
+    if (new_replication_factor < current_replication_factor) {
+        auto res = co_await decrease_replication_factor(
+          topic, new_replication_factor);
+
+        if (!res.has_value()) {
+            co_return errc::invalid_request;
+        }
+        new_replicas = res.value();
+    } else {
+        auto res = co_await increase_replication_factor(
+          topic, new_replication_factor, allocation_units);
+
+        if (!res.has_value()) {
+            co_return errc::invalid_request;
+        }
+        new_replicas = res.value();
+    }
+
+    move_topic_replicas_cmd cmd(topic, std::move(new_replicas));
+
+    co_return co_await replicate_and_wait(
+      _stm, _features, _as, std::move(cmd), timeout);
+}
+
+ss::future<result<topics_frontend::new_replicas_assigment>>
+topics_frontend::increase_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor_type new_replication_factor,
+  std::optional<allocation_units>& units) {
```

Comment:
is units also an output variable? could we avoid that?

Reply:
Unit should exists during `replicate_and_wait`. So why I pass this strange optional value inside

Reply:
but this method modifies it, right?

Reply:
it's probably ok. i'll look again closer on next review